kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/PaginatedScanSuite.scala

```diff
@@ -0,0 +1,268 @@
+          .asInstanceOf[PaginatedScanImpl]
+
+      /**
+       * Creates a custom engine with batch size 4;
+       * 00000000000000000002.json contains 2 batches, 5 active AddFiles in total
+       * Batch 1: 4 rows, 3 selected AddFiles
+       * Batch 2: 2 rows, 2 selected AddFiles
+       *
+       * 00000000000000000001.json contains 2 batches, 5 active AddFiles in total
+       * * Batch 1: 4 rows, 3 selected AddFiles
+       * * Batch 2: 2 rows, 2 selected AddFiles
+       *
+       * 00000000000000000000.json contains 2 batches, 5 active AddFiles in total
+       * * Batch 1: 4 rows, 3 selected AddFiles
+       * * Batch 2: 2 rows, 2 selected AddFiles
+       */
+      val hadoopConf = new org.apache.hadoop.conf.Configuration()
+      hadoopConf.set("delta.kernel.default.json.reader.batch-size", "4")
+      hadoopConf.set("delta.kernel.default.parquet.reader.batch-size", "4")
+      val customEngine = DefaultEngine.create(hadoopConf)
+
+      val paginatedIter = paginatedScan.getScanFiles(customEngine)
+      val firstPageFiles = collection.mutable.Buffer[FilteredColumnarBatch]()
+
+      while (paginatedIter.hasNext) {
+        val batch = paginatedIter.next()
+        firstPageFiles += batch
+      }
+      assert(firstPageFiles.nonEmpty, "First page should contain some files")
+
+      /** Verify we get 4 batches and 10 ScanFiles in total in the first page */
+      val fileCounts: Seq[Long] = firstPageFiles.map(_.getPreComputedNumSelectedRows.get().toLong)
+      val totalFileCountsReturned = fileCounts.sum
+      assert(
+        totalFileCountsReturned == 10,
+        s"First page should contain 10 files, got $totalFileCountsReturned")
+
+      assert(
+        fileCounts.length == 4,
+        s"First page should contain 3 batches, got ${fileCounts.length}")
+
+      logger.info(s"Total num batches returned in page one = ${fileCounts.length}")
+      logger.info(s"Total num Parquet Files fetched in page one = $totalFileCountsReturned")
+
+      /** Verify we get correct page token for next page */
+      val lastReadLogFileName = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+        .getLastReadLogFileName
+      val lastReturnedRowIndex = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+        .getLastReturnedRowIndex
+
+      assert(
+        lastReadLogFileName.endsWith("00000000000000000001.json"),
+        s"New Page Token: LastReadLogFilename should be 00000000000000000001.json," +
+          s"got $lastReadLogFileName")
+
+      assert(
+        lastReturnedRowIndex == 5,
+        s"New Page Token:lastReadRowIndex should be 5, got $lastReturnedRowIndex")
+
+      logger.info(s"New PageToken: lastReadLogFileName  = $lastReadLogFileName")
+      logger.info(s"New PageToken: lastReadRowIndex = $lastReturnedRowIndex")
+
+      paginatedIter.close()
+    }
+  }
+
+  test("Read First Page: basic pagination with one checkpoint file " +
```

Comment:
I see lots of similar test setup and set validation between these three tests -- can we considate these into helper methods?